src/main/scala/org/apache/spark/shuffle/UcxShuffleBlockResolver.scala

```diff
@@ -0,0 +1,141 @@
+
+    val dataFile = getDataFile(shuffleId, mapId)
+    val dataBackFile = new RandomAccessFile(dataFile, "rw")
+
+    if (dataBackFile.length() == 0) {
+      dataBackFile.close()
+      return
+    }
+
+    fileMappings.putIfAbsent(shuffleId, new CopyOnWriteArrayList[UcpMemory]())
+    offsetMappings.putIfAbsent(shuffleId, new CopyOnWriteArrayList[UcpMemory]())
+
+    val indexFile = getIndexFile(shuffleId, mapId)
+    val indexBackFile = new RandomAccessFile(indexFile, "rw")
+    val indexFileChannel = indexBackFile.getChannel
+    val dataFileChannel = dataBackFile.getChannel
+
+    // Memory map and register data and index file.
+    val dataFileBuffer = UnsafeUtils.mmap(dataFileChannel, 0, dataBackFile.length())
+    // TODO: update to jucx-1.8.0 and add an option of ODP registration
+    val dataMemory = ucxShuffleManager.ucxNode.getContext.registerMemory(dataFileBuffer)
+    fileMappings(shuffleId).add(dataMemory)
+
+    val offsetBuffer = UnsafeUtils.mmap(indexFileChannel, 0, indexBackFile.length())
+    val offsetMemory = ucxShuffleManager.ucxNode.getContext.registerMemory(offsetBuffer)
+    offsetMappings(shuffleId).add(offsetMemory)
+
+    dataFileChannel.close()
+    dataBackFile.close()
+    indexFileChannel.close()
+    indexBackFile.close()
+
+    val fileMemoryRkey = dataMemory.getRemoteKeyBuffer
+    val offsetRkey = offsetMemory.getRemoteKeyBuffer
+
+    val metadataRegisteredMemory = memPool.get(
+      fileMemoryRkey.capacity() + offsetRkey.capacity() + 24)
+    val metadataBuffer = metadataRegisteredMemory.getBuffer.slice()
+
+    if (metadataBuffer.remaining() > ucxShuffleManager.ucxShuffleConf.metadataBlockSize) {
+      throw new SparkException(s"Metadata block size ${metadataBuffer.remaining()} " +
+        s"is greater then configured 2 * ${ucxShuffleManager.ucxShuffleConf.RKEY_SIZE.key}" +
+        s"(${ucxShuffleManager.ucxShuffleConf.metadataBlockSize}).")
+    }
+
+    metadataBuffer.clear()
+
+    metadataBuffer.putLong(offsetMemory.getAddress)
+    metadataBuffer.putLong(dataMemory.getAddress)
+
+    metadataBuffer.putInt(offsetRkey.capacity())
+    metadataBuffer.putInt(fileMemoryRkey.capacity())
+
+    metadataBuffer.put(offsetRkey)
+    metadataBuffer.put(fileMemoryRkey)
+
+    metadataBuffer.clear()
+
+    val workerWrapper = ucxShuffleManager.ucxNode.getThreadLocalWorker
+    val driverMetadata = workerWrapper.getDriverMetadataBuffer(shuffleId)
+    val driverOffset = driverMetadata.address +
+      mapId * ucxShuffleManager.ucxShuffleConf.metadataBlockSize
+
+    val driverEndpoint = workerWrapper.driverEndpoint
+    val request = driverEndpoint.putNonBlocking(UcxUtils.getAddress(metadataBuffer),
+      metadataBuffer.remaining(), driverOffset, driverMetadata.ucpRkey, null)
+
+    workerWrapper.preconnnect()
+    // Blocking progress needed to make sure last mapper published data to driver before
+    // reducer starts.
+    workerWrapper.progressRequest(request)
```

Comment:
unrelated: i think we need to rename this function in JUCX to waitForRequest
"progress" sounds like it is non-blocking